cairo/ethereum/cancun/vm/instructions/system.cairo

```diff
@@ -1,58 +1,253 @@
+    with memory {
+        let calldata = memory_read_bytes(memory_input_start_position, memory_input_size);
+    }
+    EvmImpl.set_memory(memory);
+
+    let env = evm.value.env;
+    let state = env.value.state;
+    let account = get_account{state=state}(code_address);
+
+    EnvImpl.set_state{env=env}(state);
+    let env = env;
+    EvmImpl.set_env(env);
+    let evm = evm;
+
+    let code = account.value.code;
+
+    if (is_staticcall.value != 0) {
+        tempvar is_static = bool(1);
+    } else {
+        tempvar is_static = evm.value.message.value.is_static;
+    }
+    let is_static = is_static;
+
+    // TODO: this could be optimized using a non-copy mechanism.
+    let (accessed_addresses_copy_start, accessed_addresses_copy) = dict_copy(
+        cast(evm.value.accessed_addresses.value.dict_ptr_start, DictAccess*),
+        cast(evm.value.accessed_addresses.value.dict_ptr, DictAccess*),
+    );
+    let (accessed_storage_keys_copy_start, accessed_storage_keys_copy) = dict_copy(
+        cast(evm.value.accessed_storage_keys.value.dict_ptr_start, DictAccess*),
+        cast(evm.value.accessed_storage_keys.value.dict_ptr, DictAccess*),
+    );
+    tempvar child_accessed_addresses = SetAddress(
+        new SetAddressStruct(
+            cast(accessed_addresses_copy_start, SetAddressDictAccess*),
+            cast(accessed_addresses_copy, SetAddressDictAccess*),
+        ),
+    );
+    tempvar child_accessed_storage_keys = SetTupleAddressBytes32(
+        new SetTupleAddressBytes32Struct(
+            cast(accessed_storage_keys_copy_start, SetTupleAddressBytes32DictAccess*),
+            cast(accessed_storage_keys_copy, SetTupleAddressBytes32DictAccess*),
+        ),
+    );
+
+    tempvar child_message = Message(
+        new MessageStruct(
+            caller=caller,
+            target=To(new ToStruct(cast(0, Bytes0*), &to)),
+            current_target=to,
+            gas=gas,
+            value=value,
+            data=calldata,
+            code_address=&code_address,
+            code=code,
+            depth=Uint(evm.value.message.value.depth.value + 1),
+            should_transfer_value=should_transfer_value,
+            is_static=is_static,
+            accessed_addresses=child_accessed_addresses,
+            accessed_storage_keys=child_accessed_storage_keys,
+            parent_evm=evm,
+        ),
+    );
+
+    // prepare arguments to jump to process_message
+    [ap] = range_check_ptr, ap++;
+    [ap] = bitwise_ptr, ap++;
+    [ap] = keccak_ptr, ap++;
+    [ap] = poseidon_ptr, ap++;
+    [ap] = child_message.value, ap++;
+    [ap] = env.value, ap++;
+
+    call abs process_message_label;
+
+    let range_check_ptr = [ap - 6];
+    let bitwise_ptr = cast([ap - 5], BitwiseBuiltin*);
+    let keccak_ptr = cast([ap - 4], KeccakBuiltin*);
+    let poseidon_ptr = cast([ap - 3], PoseidonBuiltin*);
+    let child_evm_ = cast([ap - 2], EvmStruct*);
+    let err = cast([ap - 1], EthereumException*);
```

Comment:
after rebasing this will no longer return an err, dont forge tto update ap offsets

Reply:
so below we can't return early anymore, right?

Reply:
yes, we dont need to